backend/index.js

```diff
@@ -1,283 +1,374 @@
-    for (let childId of this.children[objectId]) {
-      this.makePatch(childId, diffs)
+    // Detect duplicate assignments to the same object and key
+    if (['set', 'del', 'link', 'inc'].includes(op.action) && !op.insert) {
+      if (!assignments[op.obj]) {
+        assignments[op.obj] = {[op.key]: op}
+      } else if (!assignments[op.obj][op.key]) {
+        assignments[op.obj][op.key] = op
+      } else if (op.action === 'inc') {
+        assignments[op.obj][op.key].value += op.value
+        continue
+      } else {
+        assignments[op.obj][op.key].action = op.action
+        assignments[op.obj][op.key].value = op.value
+        continue
+      }
     }
-    for (let d of this.diffs[objectId]) diffs.push(d)
+
+    change.ops.push(op)
   }
-}
 
+  return [state.set('objectIds', objectIds), change]
+}
 
 /**
  * Returns an empty node state.
  */
 function init() {
-  return Map({opSet: OpSet.init()})
+  const opSet = OpSet.init(), versionObj = Map({version: 0, localOnly: true, opSet})
+  const state = Map({opSet, versions: List.of(versionObj), objectIds: Map()})
+  return {state}
+}
+
+function clone(backend) {
+  return {state: backendState(backend)}
+}
+
+function free(backend) {
+  backend.state = null
+  backend.frozen = true
 }
 
 /**
- * Constructs a patch object from the current node state `state` and the list
- * of object modifications `diffs`.
+ * Constructs a patch object from the current node state `state` and the
+ * object modifications `diffs`.
  */
-function makePatch(state, diffs) {
+function makePatch(state, diffs, request, isIncremental) {
+  const version = state.get('versions').last().get('version')
+  const clock   = state.getIn(['opSet', 'states']).map(seqs => seqs.size).toJSON()
+  const deps    = state.getIn(['opSet', 'deps']).toJSON().sort()
   const canUndo = state.getIn(['opSet', 'undoPos']) > 0
   const canRedo = !state.getIn(['opSet', 'redoStack']).isEmpty()
-  const clock = state.getIn(['opSet', 'clock']).toJS()
-  const deps = state.getIn(['opSet', 'deps']).toJS()
-  return {clock, deps, canUndo, canRedo, diffs}
+  const patch = {version, clock, deps, canUndo, canRedo, diffs}
+
+  if (isIncremental && request) {
+    patch.actor = request.actor
+    patch.seq   = request.seq
+  }
+  return patch
 }
 
 /**
- * The implementation behind `applyChanges()` and `applyLocalChange()`.
+ * The implementation behind `applyChanges()`, `applyLocalChange()`, and
+ * `loadChanges()`.
  */
-function apply(state, changes, undoable) {
-  let diffs = [], opSet = state.get('opSet')
-  for (let change of fromJS(changes)) {
-    change = change.remove('requestType')
-    const [newOpSet, diff] = OpSet.addChange(opSet, change, undoable)
-    for (let d of diff) diffs.push(d)
-    opSet = newOpSet
+function apply(state, changes, request, isUndoable, isIncremental) {
+  let diffs = isIncremental ? {} : null
+  let opSet = state.get('opSet')
+  for (let change of changes) {
+    for (let chunk of splitContainers(change)) {
```

Comment:
`chunk` is never used inside the loop... does this only work because the changes only contain one chunk by accident?

Reply:
You're right, it should be using `chunk` instead of `change` inside the loop. I've pushed e3d0170b1249ed401cf38713776925aad981e6e9 to fix this. Thanks for pointing it out!